File: 101. Symmetric Tree/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isSymmetric = function(root) {
    // mid order traverse cannot be used
    return isMirror(root, root)
};
    
var isMirror = function(p, q) {
    if(!p && !q) return true
    if(!p && q) return false
    if(p && !q) return false
    if(p.val !== q.val) return false
    return isMirror(p.left, q.right) && isMirror(p.right, q.left) // compare left part with right part
}

File: 101. Symmetric Tree/solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSymmetric(root *TreeNode) bool {
    if root == nil {
        return true
    }
    return isSymmetricSameTree(root.Left, root.Right) 
}

func isSymmetricSameTree(p *TreeNode, q *TreeNode) bool {
    if p == nil && q == nil {   // important
        return true
    }
    if p == nil && q != nil {   // important
        return false
    }
    if p != nil && q == nil {   // important
        return false
    }
    return p.Val == q.Val && isSymmetricSameTree(p.Left, q.Right) && isSymmetricSameTree(p.Right, q.Left)
}